bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/TypeChecker.java

```
@@ -1295,4 +1298,127 @@ public boolean equals(Object otherPair) {
+        if (type.isNullable()) {
+            return true;
+        }
+        // All members are of same type.
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember;
+        for (firstMember = iterator.next(); iterator.hasNext(); ) {
+            if (!isSameType(firstMember, iterator.next())) {
+                return false;
+            }
+        }
+        // Control reaching this point means there is only one type in the union.
+        return BTypes.isValueType(firstMember) && hasImplicitInitialValue(firstMember);
+    }
+
+    private static boolean analyzeRecordType(BStructureType type, List<BType> unanalyzedTypes) {
+        if (unanalyzedTypes.contains(type)) {
+            return true;
+        }
+        unanalyzedTypes.add(type);
+        return type.getFields().values().stream().allMatch(f -> hasImplicitInitialValue(f.type, unanalyzedTypes));
+    }
+
+    private static boolean analyzeObjectType(BObjectType type) {
+        if (type.getTag() == TypeTags.SERVICE_TAG) {
+            return false;
+        }
+        if (type.getTag() == TypeTags.OBJECT_TYPE_TAG) {
+            AttachedFunction initializerFunc = type.initializer;
+            if (initializerFunc == null) {
+                // No __init function found.
+                return true;
+            }
+            BFunctionType initFuncType = initializerFunc.type;
+            boolean noParams = initFuncType.paramTypes.length == 0;
+            boolean nilReturn = initFuncType.retType.getTag() == TypeTags.NULL_TAG;
+            return noParams && nilReturn;
+        }
+        return false;
+    }
+
+    private static boolean analyzeFiniteType(BFiniteType type) {
+        // Has NIL element as a member.
+        if (type.valueSpace.stream().anyMatch(Objects::isNull)) {
+            return true;
+        }
+        // All of the element are from same type. And elements contains implicit initial value.
+        if (type.valueSpace.isEmpty()) {
```

Is this ever satisfied? Also not sure if the comment is valid?

---

At runtime, no.. There wouldn't be empty value spaces. But even in compile time, this should've thrown an exception, instead of simply returning false IMO.